src/provider/provider_devdax_memory.c

```diff
@@ -0,0 +1,655 @@
+ * SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+*/
+
+#include <assert.h>
+#include <errno.h>
+#include <limits.h>
+#include <stddef.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+
+#include "base_alloc_global.h"
+#include "critnib.h"
+#include "provider_devdax_memory_internal.h"
+#include "provider_os_memory_internal.h"
+#include "utils_common.h"
+#include "utils_concurrency.h"
+#include "utils_log.h"
+
+#include <umf.h>
+#include <umf/memory_provider_ops.h>
+#include <umf/providers/provider_devdax_memory.h>
+
+#define NODESET_STR_BUF_LEN 1024
+
+#define TLS_MSG_BUF_LEN 1024
+
+typedef struct devdax_last_native_error_t {
+    int32_t native_error;
+    int errno_value;
+    char msg_buff[TLS_MSG_BUF_LEN];
+} devdax_last_native_error_t;
+
+static __TLS devdax_last_native_error_t TLS_last_native_error;
+
+// helper values used only in the Native_error_str array
+#define _UMF_DEVDAX_RESULT_SUCCESS                                             \
+    (UMF_DEVDAX_RESULT_SUCCESS - UMF_DEVDAX_RESULT_SUCCESS)
+#define _UMF_DEVDAX_RESULT_ERROR_ALLOC_FAILED                                  \
+    (UMF_DEVDAX_RESULT_ERROR_ALLOC_FAILED - UMF_DEVDAX_RESULT_SUCCESS)
+#define _UMF_DEVDAX_RESULT_ERROR_ADDRESS_NOT_ALIGNED                           \
+    (UMF_DEVDAX_RESULT_ERROR_ADDRESS_NOT_ALIGNED - UMF_DEVDAX_RESULT_SUCCESS)
+#define _UMF_DEVDAX_RESULT_ERROR_FREE_FAILED                                   \
+    (UMF_DEVDAX_RESULT_ERROR_FREE_FAILED - UMF_DEVDAX_RESULT_SUCCESS)
+#define _UMF_DEVDAX_RESULT_ERROR_PURGE_FORCE_FAILED                            \
+    (UMF_DEVDAX_RESULT_ERROR_PURGE_FORCE_FAILED - UMF_DEVDAX_RESULT_SUCCESS)
+
+static const char *Native_error_str[] = {
+    [_UMF_DEVDAX_RESULT_SUCCESS] = "success",
+    [_UMF_DEVDAX_RESULT_ERROR_ALLOC_FAILED] = "memory allocation failed",
+    [_UMF_DEVDAX_RESULT_ERROR_ADDRESS_NOT_ALIGNED] =
+        "allocated address is not aligned",
+    [_UMF_DEVDAX_RESULT_ERROR_FREE_FAILED] = "memory deallocation failed",
+    [_UMF_DEVDAX_RESULT_ERROR_PURGE_FORCE_FAILED] = "force purging failed",
+};
+
+static void devdax_store_last_native_error(int32_t native_error,
+                                           int errno_value) {
+    TLS_last_native_error.native_error = native_error;
+    TLS_last_native_error.errno_value = errno_value;
+}
+
+static umf_result_t
+devdax_translate_params(umf_devdax_memory_provider_params_t *in_params,
+                        devdax_memory_provider_t *provider) {
+    umf_result_t result;
+
+    result = os_translate_mem_protection_flags(in_params->protection,
+                                               &provider->protection);
+    if (result != UMF_RESULT_SUCCESS) {
+        LOG_ERR("incorrect memory protection flags: %u", in_params->protection);
+        return result;
+    }
+
+    /* devdax requires UMF_MEM_MAP_SHARED */
+    result = os_translate_mem_visibility_flag(UMF_MEM_MAP_SHARED,
```

Comment:
Do we need dax/pmem mmap magic here? MAP_SYNC(and maybe MAP_SYNC_VALIDATE). 

If yes, please add it, and second more important add test enuring that we are not going thru page cache

Reply:
@lplewa mapping with `MAP_SHARED_VALIDATE | MAP_SYNC` added.

Reply:
It works:
```
32: [PID:167087 TID:167087 DEBUG UMF] os_devdax_mmap: devdax mapped with the (MAP_SHARED_VALIDATE | MAP_SYNC) flags
```

Reply:
Done

Reply:
Added the test checking if devdax was mapped with the `MAP_SYNC` flag.

Done